codegen/src/main/java/software/amazon/awssdk/codegen/poet/builder/BaseClientBuilderClass.java

```diff
@@ -180,27 +176,45 @@ private MethodSpec beanStyleSetServiceConfigurationMethod() {
                          .build();
     }
 
-    private MethodSpec serviceSpecificHttpConfigMethod() {
-        return MethodSpec.methodBuilder("serviceHttpConfig")
-                         .addAnnotation(Override.class)
-                         .addModifiers(Modifier.PROTECTED, Modifier.FINAL)
-                         .returns(AttributeMap.class)
-                         .addCode("return $T.defaultHttpConfig();",
-                                  PoetUtils.classNameFromFqcn(model.getCustomizationConfig().getServiceSpecificHttpConfig()))
-                         .build();
+    private void addServiceHttpConfigIfNeeded(TypeSpec.Builder builder, IntermediateModel model) {
+        String serviceDefaultFqcn = model.getCustomizationConfig().getServiceSpecificHttpConfig();
+        boolean supportsH2 = model.getMetadata().supportsH2();
+
+        if (serviceDefaultFqcn != null || supportsH2) {
+            builder.addMethod(serviceSpecificHttpConfigMethod(serviceDefaultFqcn, supportsH2));
+        }
     }
 
-    private MethodSpec enableH2HttpConfigMethod() {
+    private MethodSpec serviceSpecificHttpConfigMethod(String serviceDefaultFqcn, boolean supportsH2) {
         return MethodSpec.methodBuilder("serviceHttpConfig")
                          .addAnnotation(Override.class)
                          .addModifiers(Modifier.PROTECTED, Modifier.FINAL)
                          .returns(AttributeMap.class)
-                         .addCode("return $T.builder()\n"
-                                  + ".put($T.PROTOCOL, $T.HTTP2)\n"
-                                  + ".build();", AttributeMap.class, SdkHttpConfigurationOption.class, Protocol.class)
+                         .addCode(serviceSpecificHttpConfigMethodBody(serviceDefaultFqcn, supportsH2))
                          .build();
     }
 
+    private CodeBlock serviceSpecificHttpConfigMethodBody(String serviceDefaultFqcn, boolean supportsH2) {
+        CodeBlock.Builder builder =  CodeBlock.builder()
+                                              .addStatement("$1T result = $1T.builder().build()", AttributeMap.class);
+
+        if (serviceDefaultFqcn != null) {
+            builder.addStatement("result = result.merge($T.defaultHttpConfig())",
+                                 PoetUtils.classNameFromFqcn(model.getCustomizationConfig().getServiceSpecificHttpConfig()));
+        }
+
+        if (supportsH2) {
```

Comment:
Minor but can we move this up so we only have two merges instead of three when h2 and service specific config is present? Also would avoid mutation of a local var. I.E. code that looks like this

```java
AttributeMap result = AttributeMap.empty();
return result.merge(serviceConfig);
```

or 

```java
AttributeMap result = AttributeMap.builder()
    .put(PROTOCOL, HTTP2)
    .build();
return result.merge(serviceConfig);
```

Reply:
Updated